Added Inner Classes to Class Naming Conventions Rule

sgnl-labs opened the pull request:
## Describe the PR

Added inner classes to ClassNamingConventionsRule, and added property to support name checking for inner classes.

## Related issues

None (this was uncovered while developing custom rules for Apex

## Ready?

<!-- If you feel like you can help to check off the following tasks, that'd be great. If not, don't worry - we will take care of it. -->

- [ ] Added unit tests for fixed bug/feature
- [ ] Passing all unit tests
- [ ] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [ ] Added (in-code) documentation (if needed)

adangel reviewed (changes requested):
Thanks for the PR!

Could you please add additional test cases in https://github.com/pmd/pmd/blob/master/pmd-apex/src/test/resources/net/sourceforge/pmd/lang/apex/rule/codestyle/xml/ClassNamingConventions.xml
See https://docs.pmd-code.org/latest/pmd_userdocs_extending_testing.html for more information about rule testing.

Also I would be interested in:
* Is this common for Apex, that inner classes / interfaces are created?
* Can you give an example of the patterns, you would use?

sgnl-labs commented:
> Thanks for the PR!
> 
> Could you please add additional test cases in https://github.com/pmd/pmd/blob/master/pmd-apex/src/test/resources/net/sourceforge/pmd/lang/apex/rule/codestyle/xml/ClassNamingConventions.xml See https://docs.pmd-code.org/latest/pmd_userdocs_extending_testing.html for more information about rule testing.
> 
> Also I would be interested in:
> 
> * Is this common for Apex, that inner classes / interfaces are created?

I wouldn't say that it's a common practice, but it is out there. There are some examples in the official API, but essentially it boils down to organization and consistency. It can also be used to isolate classes that have hidden functionality (although this can be done in other ways, inner classes have the advantage of readability.) One example is `Schema.SObjectType`. `Schema `is a class, but so is `SObjectType`. It just provides organization.

> * Can you give an example of the patterns, you would use?

Most recently, a platform cached provider of RecordType objects grouped by Salesforce [SObject ](url)(meaning database table, essentially an ORM object.) The 'provider/repository' class was the outer class. The classes that applied directly to the object (but inherited from another, separate class) were inner classes.

Although there isn't really any functional benefit to inner classes, they provide organization on a platform that doesn't allow Java-like namespaces in its Apex source code.

...and thanks for allowing me to possibly contribute to this amazing project! I have so many questions for you guys, but I wanted to start with this small offering. The reason is simple: in a project I was working on, we used class names as 'namespace indicators'...a domain prefix if you will. They had a specific standard that was enforced by PMD. Unfortunately, inner classes were being held to the same standard as outer classes and that didn't align with the standard we wanted to follow.

sgnl-labs commented:
> * Can you give an example of the patterns, you would use?

A colleague of mine put it much more succinctly: 

> In MVC design terminology, it is useful to define an inner class to define the data structure being returned to the view & containing its scope to the controller which interacts with the model. It provides containment/encapsulation.

adangel commented:
> They had a specific standard that was enforced by PMD. Unfortunately, inner classes were being held to the same standard as outer classes and that didn't align with the standard we wanted to follow.

Now I understand the problem, thanks for clarifying. In that case, adding a separate property for inner class regex is probably the most flexible solution. Another way could be, that inner classes/interfaces are simply not considered by the rule - but then, you won't be able to enforce any naming conventions on them.

sgnl-labs commented:
> > They had a specific standard that was enforced by PMD. Unfortunately, inner classes were being held to the same standard as outer classes and that didn't align with the standard we wanted to follow.
> 
> Now I understand the problem, thanks for clarifying. In that case, adding a separate property for inner class regex is probably the most flexible solution. Another way could be, that inner classes/interfaces are simply not considered by the rule - but then, you won't be able to enforce any naming conventions on them.

The property solution is definitely the most flexible path. I already implemented it like that because the rule I modified was doing that anyways. I did toy around with one other possibility:

Summit (the Apex parser PMD uses now) already knows how to make the distinction between inner and outer classes, so why not make inner classes part of the `ASTUserClass` or `ASTUserClassOrInterface` object(s)? I tossed those ideas because they seemed too intrusive and I believed that to be a risk to overall design hygiene.

@adangel I am curious about something too, because I frequently find myself buried 'under the hood' of PMD for custom rules. Is there a reason that ANTLR4 isn't a 'strongly suggested' standard for every language? I was looking at different language modules and it seems to be kind of the Wild West.

sgnl-labs commented:
So I've been thinking about it, and this should cover a much broader set of scenarios than just naming conventions via a rule. In fact, exposing inner classes, interfaces, and enums via the `ASTUserClassOrInterface` at the parent level would allow the inner/outer hierarchy to be leveraged by any rule.

Any thoughts on that @adangel ?

adangel reviewed (changes requested):
Thanks for creating the test cases. Unfortunately it fails at the moment already before executing the tests with our PMD checks, see the comments.

> Summit already knows how to make the distinction between inner and outer classes, so why not make inner classes part of the ASTUserClass or ASTUserClassOrInterface object(s)?

The inner classes are already part of the outer classes in the tree, that's why you can figure it out via `getParent()`.

> In fact, exposing inner classes, interfaces, and enums via the ASTUserClassOrInterface at the parent level would allow the inner/outer hierarchy to be leveraged by any rule.

That is possible, e.g. create a default method in ASTUserClassOrInterface named `isNested()` - then you can simply use this method in the rule. See https://github.com/pmd/pmd/blob/e6397d5d5ce613d57280c04bf97cddd679259e09/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTTypeDeclaration.java#L189-L195 for the equivalent implementation for Java.

> @adangel I am curious about something too, because I frequently find myself buried 'under the hood' of PMD for custom rules. Is there a reason that ANTLR4 isn't a 'strongly suggested' standard for every language? I was looking at different language modules and it seems to be kind of the Wild West.

PMD started out with only supporting JavaCC grammars, since PMD 7, we also support Antlr. In addition, we also support preexisting ASTs (independent of any grammar), that need to be mapped (like Apex Summit-AST, JavaScript).
See https://docs.pmd-code.org/latest/pmd_devdocs_major_adding_new_language_antlr.html and https://docs.pmd-code.org/latest/pmd_devdocs_major_adding_new_language_javacc.html - these pages describe what it means to support a language in PMD...

> (FYI: sgnl-labs and justinstroudbah are one and the same person...for whatever reason my initial commit was filed under the account I use for my day job.)

This is about the user name/email configuration in git, see https://git-scm.com/docs/git-config#Documentation/git-config.txt-username
GitHub maps the email to a user account then for attribution.


Another point: do we need a separate pattern for inner (nested) enums? And which pattern should be applied for a inner, abstract class? I actually don't want to add too many patterns. So, maybe we can ignore these cases for now until they are really needed?

sgnl-labs reviewed (commented):
Committed suggested changes, thank you for the input!

sgnl-labs commented:
> Another point: do we need a separate pattern for inner (nested) enums? And which pattern should be applied for a inner, abstract class? I actually don't want to add too many patterns. So, maybe we can ignore these cases for now until they are really needed?

@adangel  I think we can safely ignore those cases. Inner classes aren't used heavily, and to your point introducing too much 'predicted complexity' spoils the sauce, so to speak. The main goal I had when introducing this was just to give flexibility in naming conventions.

adangel reviewed (approved):
Thanks!

I've added also "innerInterfacePatterns" to the test cases and introduced `isNested()`.

I'll merge it soon (once the build is successful).

adangel commented:
> Inner enums are actually more common than inner classes.

So, currently the rule can't distinguish between an outer enum and an inner enum. If this is wanted, a new property is needed. I'd suggest to do this then in a new PR, if needed.

pmd-test commented:
<!--
  0 Errors
  0 Warnings
  1 Message: Compared to master:<br />
This...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to master:<br />
This changeset changes 0 violations,<br />
introduces 0 new violations, 0 new errors and 0 new configuration errors,<br />
removes 0 violations, 0 errors and 0 configuration errors.<br />
<a href="https://github.com/pmd/pmd/actions/runs/9988911233?pr=5048">Download full report as build artifact</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

sgnl-labs commented:
> > Inner enums are actually more common than inner classes.
> 
> So, currently the rule can't distinguish between an outer enum and an inner enum. If this is wanted, a new property is needed. I'd suggest to do this then in a new PR, if needed.

Understood. I'll take a look this weekend (we're in a release cycle right now so my time isn't as free as I'd like!) Since they are more common than inner classes, I think this makes sense.

sgnl-labs commented:
Also, @adangel, does the project make a habit of deleting old branches? Just asking as I don't want to leave a messy campsite so to speak.

adangel commented:
@sgnl-labs You can safely your branch (it is only in your fork anyway). We still have this PR in case we need to look at the changes in context.
